internal/handlers/commonpath/commonpath.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package commonpath contains functions used for path.
+package commonpath
+
+import (
+	"errors"
+	"strconv"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/iterator"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// FindValuesOpts sets options for FindValues.
+type FindValuesOpts struct {
+	// SearchInArray searches by iterating through the whole array to find documents that contains path key.
+	// Using path `v.foo` and `v` is an array, it returns all document which has key `foo`.
+	// If `v` is not an array, SearchInArray has no impact.
+	SearchInArray bool
+	// FindArrayIndex gets an element at the specified index of an array.
+	// Using path `v.0` and `v` is an array, it returns 0-th index element of the array.
+	// If `v` is not an array, FindArrayIndex has no impact.
+	FindArrayIndex bool
+}
+
+// FindValues goes through each key of the path iteratively on doc to find values
+// at the suffix of the path. At each key of the path, it checks:
+//   - if the document has the key;
+//   - if the array contains an element at an index where index equals to the key, and;
+//   - if the array contains one or more documents which have the key.
```

Comment:
I don't think I understand those two lines completely, but they are more clean the option names and descriptions above. It is hard to understand what options match what line there.

Reply:
I added which option is used, hopefully it's better :sweat_smile: